core/src/inputs/advice/merkle_path_set.rs

```diff
@@ -0,0 +1,318 @@
+    /// # Errors
+    /// Returns an error if:
+    /// * The specified index not valid for the depth of structure.
+    pub fn get_path(&self, depth: u32, index: u64) -> Result<Vec<Word>, AdviceSetError> {
+        if index >= 2u64.pow(self.total_depth) {
+            return Err(AdviceSetError::InvalidIndex(self.total_depth, index));
+        }
+        if depth != self.total_depth {
+            return Err(AdviceSetError::InvalidDepth(self.total_depth, depth));
+        }
+
+        let pos = 2u64.pow(depth) + index;
+        let index = pos / 2;
+
+        let mut path = self.paths[&index].clone();
+
+        if is_even(pos) {
+            path.remove(0);
+            Ok(path)
+        } else {
+            path.remove(1);
+            Ok(path)
+        }
+    }
+
+    /// Replaces the leaf at the specified index with the provided value.
+    ///
+    /// # Errors
+    /// Returns an error if the node with specified index is not in the map.
+    pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<(), AdviceSetError> {
+        let depth = self.depth();
+        if index >= 2u64.pow(depth) {
+            return Err(AdviceSetError::InvalidIndex(depth, index));
+        }
+
+        // Fill old_hashes vector -----------------------------------------------------------------
+        let pos = 2u64.pow(depth) + index;
+        let path = self.paths.get(&(pos / 2)).unwrap().clone();
+
+        let (old_hashes, _) = compute_hashes(&path, depth, index);
+
+        // Fill new_hashes vector -----------------------------------------------------------------
+        let pos = 2u64.pow(depth) + index;
+
+        if is_even(pos) {
+            self.paths.get_mut(&(pos / 2)).unwrap()[0] = value;
+        } else {
+            self.paths.get_mut(&(pos / 2)).unwrap()[1] = value;
+        }
+
+        let path = self.paths.get(&(pos / 2)).unwrap().clone();
+
+        let (new_hashes, new_root) = compute_hashes(&path, depth, index);
+        self.root = new_root;
+
+        // update paths ---------------------------------------------------------------------------
+        for path in self.paths.values_mut() {
+            for i in (0..old_hashes.len()).rev() {
+                if path[i + 2] == old_hashes[i] {
+                    path[i + 2] = new_hashes[i];
+                    break;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// HELPER FUNCTIONS
+// --------------------------------------------------------------------------------------------
+
+fn is_even(pos: u64) -> bool {
+    pos & 1 == 0
+}
+
+// pos -- position of the cur hash
+pub fn calculate_parent_hash(cur: Word, pos: u64, hash_2: Word) -> Word {
+    if is_even(pos) {
+        hasher::merge(&[cur.into(), hash_2.into()]).into()
+    } else {
+        hasher::merge(&[hash_2.into(), cur.into()]).into()
+    }
+}
+
+/// Returns vector of hashes from current to the root
+pub fn compute_hashes(path: &[Word], depth: u32, index: u64) -> (Vec<Word>, Word) {
```

Comment:
nit: let's rename this to `compute_path_trace`. Also, I don't think this function needs to be public.